Allow tensor subclasses and add `torch.serialization.add_safe_globals` that allows users to allowlist classes for `weights_only` load

mikaylagawarecki opened the pull request:
#### Conditions for allowlisting tensor subclasses
We allow tensor subclasses types that 
(1) Do not override `__setstate__`, `__getattr__`, `__setattr__`, `__get__`, `__set__` or `__getattribute__` of `torch.Tensor` (`torch.Tensor` does not have a definition of `__getattr__`, `__get__` or `__set__` so we check that these are `None`)
(2) Use the generic `tp_alloc`
(3) Are in a module that *has been imported by the user*
to be pushed onto the stack as strings by `GLOBAL` instructions, while storing the type in a dict

The strings will be converted to the classes as appropriate when executing `REBUILD` with `_rebuild_from_type_v2`

*Note that we use `inspect.getattr_static(sys.modules[module], name)` to get the class/function as this method claims to have no code execution.

The rationale for the 3 conditions above is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `getattr`, `Tensor.__setstate__` and the call to `as_subclass` as well as the call to `_set_obj_state` which calls `setattr`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.{add/get/clear}_safe_globals` that enables user to allowlist globals they have deemed safe and manipulate this list (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe).

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124331

mikaylagawarecki committed:
Add API for users to allowlist classes for weights_only for Tensor Subclasses

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add API for users to allowlist classes for weights_only for Tensor Subclasses"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

Allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) and added `torch.serialization.mark_safe_globals` that enables user allowlisting of classes they have deemed safe + simple test with `TwoTensor`.

This API is primarily intended to be used for checkpoints where user-defined Tensor Subclasses have been serialized.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki reviewed (commented):
Marking as draft after discussion with Alban -- while we will likely still need a way for users to allowlist specific classes, want to also look into whether there is a generic way to allowlist tensor subclasses (with appropriate guardrails for security)

mikaylagawarecki committed:
Update on "Add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

Allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) and added `torch.serialization.mark_safe_globals` that enables user allowlisting of classes they have deemed safe + simple test with `TwoTensor`.

This API is primarily intended to be used for checkpoints where user-defined Tensor Subclasses have been serialized.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +176,31 @@ def load(self):
+                    self.append(_get_user_allowed_globals()[full_path])
                 else:
-                    raise RuntimeError(f"Unsupported class {full_path}")
+                    class_type = getattr(modules[module], name)
```
@albanD I think there might be a fundamental problem here that prevents us from allowlisting *all* tensor subclasses.

The `GLOBAL <bla>` in the pickle file will have <bla> as a string, in order to check `issubclass(bla, torch.Tensor), we need to convert the string `bla` to the actual type object. To my knowledge, I could not find a safe way to do this, and I suspect this line here might not be what we consider secure. wdyt?

Is there a way where we can constrain this such that it is still secure?

mikaylagawarecki committed:
Update on "Add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

Allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) and added `torch.serialization.mark_safe_globals` that enables user allowlisting of classes they have deemed safe + simple test with `TwoTensor`.

This API is primarily intended to be used for checkpoints where user-defined Tensor Subclasses have been serialized.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

Allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) and added `torch.serialization.mark_safe_globals` that enables user allowlisting of classes they have deemed safe + simple test with `TwoTensor`.

This API is primarily intended to be used for checkpoints where user-defined Tensor Subclasses have been serialized.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allowlist tensor subclasses that 
(1) Do not override `__setstate__`
(2) Use the generic `tp_alloc`

The rationale for these two conditions is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `Tensor.__setstate__` and the call to `as_subclass`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.mark_safe_globals` that enables user to allowlist globals they have deemed safe (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allowlist tensor subclasses that 
(1) Do not override `__setstate__`
(2) Use the generic `tp_alloc`

The rationale for these two conditions is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `Tensor.__setstate__` and the call to `as_subclass`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.mark_safe_globals` that enables user to allowlist globals they have deemed safe (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

malfet commented on test/test_serialization.py:
```diff
@@ -4097,6 +4098,54 @@ def test_empty_class_serialization(self):
+
+            # Loading tensor subclass with weights_only=True should work
+            # if __setstate__ and tp_alloc are not overriden and user has imported the subclass
+            from torch.testing._internal.two_tensor import TwoTensor
+            f.seek(0)
+            sd = torch.load(f, weights_only=True)
```
IMO this is wrong: weights_only should refuse to deserialize user imported classes unless they are allowlisted. Because almost every Python program program imports `sys`, which makes it "safe" (which is clearly unsafe) to call functions from that module)

mikaylagawarecki replied:
Hm, the implementation in the unpickler will do this check though, so arbitrary functions from sys should not be callable (unless they are tensor subclasses that do not override `__setstate__` or `tp_alloc`

```
class_type = getattr(modules[module], name)
if isinstance(class_type, type) and issubclass(
    class_type, torch.Tensor
):
    # checks __setstate__ and tp_alloc
else:
     # raise Unsupported
```

Do you think this is still unsafe?

The rationale for this was that @albanD was hoping that we could "blanket" allow tensor subclasses that fall under certain restrictions

mikaylagawarecki replied:
I verified that other methods are not called by adding a test that overrides every single method of a class and noticed  that `__setattr__` and `__getattribute__` are called as well (due to `getattr` and `setattr` calls from `rebuild_from_type_v2` and the call to `_set_obj_state` in `rebuild_from_type_v2`.

I also blocklisted other functions that could override the behavior of `getattr` and `setattr` (`__get__`, `__set__`, `__getattr__`)

malfet commented on torch/_weights_only_unpickler.py:
```diff
@@ -59,11 +61,32 @@
+_marked_safe_globals_list: List[Callable] = []
+
+
+def _mark_safe_globals(safe_globals: List[Callable]):
```
From security point-of-view I would prefer the list of safe globals to be an explicit argument to torch.load rather than a global state, as relying on global state is prone to hard-to-track security violations (somebody can mark something as safe during import of third-party library and then all users of torch will assume it's safe to load `.pth`, while in reality it is not, because someplace else global state was modified to allow that

mikaylagawarecki replied:
Sorry for not updating here. Updating this to reflect the latest view:

Passing the allowed globals as a kwarg to to `torch.load` might be prohibitive as users might not have access to the `torch.load` call (e.g. the end-user facing API might be something like `from_pretrained`. As such, having such global state is a tradeoff we might need to make. We will provide an API to get the current state of the global list/remove things from the global list.

I will update the PR to do the above if we agree on the approach in the google doc.

malfet reviewed (changes requested):
From security point-of-view `torch.serialization._mark_safe_globals([torch.foo])` API feels very unsafe, first of all, because there are no way to remove something from the list at the moment, and second because doing something like that would require one to be cognizant of all potential classes that could have been marked safe in previous function calls. How about `torch.load(f, safe_globals=[...])` which implies, `weights_only`

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allowlist tensor subclasses that 
(1) Do not override `__setstate__`
(2) Use the generic `tp_alloc`

The rationale for these two conditions is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `Tensor.__setstate__` and the call to `as_subclass`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.mark_safe_globals` that enables user to allowlist globals they have deemed safe (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allowlist tensor subclasses that 
(1) Do not override `__setstate__`, `__getattr__`, `__setattr__`
(2) Use the generic `tp_alloc`
(3) Are in a module that *has been imported by the user*

*Note that we use `inspect.getattr_static(sys.modules[module], name)` to get the class as this method claims to have no code execution.

The rationale for these two conditions is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `getattr`, `Tensor.__setstate__` and the call to `as_subclass` as well as the call to `_set_obj_state` which calls `setattr`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.mark_safe_globals` that enables user to allowlist globals they have deemed safe (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allowlist tensor subclasses that 
(1) Do not override `__setstate__`, `__getattr__`, `__setattr__`
(2) Use the generic `tp_alloc`
(3) Are in a module that *has been imported by the user*

*Note that we use `inspect.getattr_static(sys.modules[module], name)` to get the class as this method claims to have no code execution.

The rationale for these two conditions is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `getattr`, `Tensor.__setstate__` and the call to `as_subclass` as well as the call to `_set_obj_state` which calls `setattr`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.mark_safe_globals` that enables user to allowlist globals they have deemed safe (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allowlist tensor subclasses that 
(1) Do not override `__setstate__`, `__getattr__`, `__setattr__`, `__get__`, `__set__` or `__getattribute__` of `torch.Tensor` (`torch.Tensor` does not have a definition of `__getattr__`, `__get__` or `__set__` so we check that these are `None`)
(2) Use the generic `tp_alloc`
(3) Are in a module that *has been imported by the user*

*Note that we use `inspect.getattr_static(sys.modules[module], name)` to get the class/function as this method claims to have no code execution.

The rationale for these two conditions is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `getattr`, `Tensor.__setstate__` and the call to `as_subclass` as well as the call to `_set_obj_state` which calls `setattr`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.mark_safe_globals` that enables user to allowlist globals they have deemed safe (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allowlist tensor subclasses that 
(1) Do not override `__setstate__`, `__getattr__`, `__setattr__`, `__get__`, `__set__` or `__getattribute__` of `torch.Tensor` (`torch.Tensor` does not have a definition of `__getattr__`, `__get__` or `__set__` so we check that these are `None`)
(2) Use the generic `tp_alloc`
(3) Are in a module that *has been imported by the user*

*Note that we use `inspect.getattr_static(sys.modules[module], name)` to get the class/function as this method claims to have no code execution.

The rationale for these two conditions is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `getattr`, `Tensor.__setstate__` and the call to `as_subclass` as well as the call to `_set_obj_state` which calls `setattr`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.mark_safe_globals` that enables user to allowlist globals they have deemed safe (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allow tensor subclasses types that 
(1) Do not override `__setstate__`, `__getattr__`, `__setattr__`, `__get__`, `__set__` or `__getattribute__` of `torch.Tensor` (`torch.Tensor` does not have a definition of `__getattr__`, `__get__` or `__set__` so we check that these are `None`)
(2) Use the generic `tp_alloc`
(3) Are in a module that *has been imported by the user*
to be pushed onto the stack by `GLOBAL` instructions.

Further, we only allow these types to be used as the second argument of `_rebuild_from_type_v2`

*Note that we use `inspect.getattr_static(sys.modules[module], name)` to get the class/function as this method claims to have no code execution.

The rationale for the 3 conditions above is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `getattr`, `Tensor.__setstate__` and the call to `as_subclass` as well as the call to `_set_obj_state` which calls `setattr`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.{mark/get/clear}_safe_globals` that enables user to allowlist globals they have deemed safe and manipulate this list (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -162,10 +283,28 @@ def load(self):
                     raise RuntimeError(
                         f"Trying to call reduce for unrecognized function {func}"
                     )
+                # Prevent tensor subclass type that is pushed onto stack in GLOBAL from
```
We do not want to expect new/init are not changed. So we cannot allow them to be called.

mikaylagawarecki replied:
To elaborate here, we want to make sure that tensor subclass types that have been allowed by the Unpickler but are not in allowed globals are **only ever used as the second argument to `torch.Tensor._rebuild_from_type_v2`** 

We want to disallow the tensor subclass type from being pushed onto the stack as an argument for any `REBUILD` instruction, in case one of its arguments (e.g. as `func` for line 58 here) 

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

If this happens, `__init__` and `__new__` of the tensor subclass might be called as mentioned above.

albanD commented on test/test_serialization.py:
```diff
@@ -4128,6 +4147,173 @@ def test_empty_class_serialization(self):
+        return bad_func
+
+    @parametrize("wrapper", (True, False))
+    def test_tensor_subclass_method_spoofing(self, wrapper):
```
Discussed offline for a more detailed comment on all the things tested here.

mikaylagawarecki commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +193,87 @@ def load(self):
                 else:
-                    raise RuntimeError(f"Unsupported class {full_path}")
+                    # For tensor subclasses.
+                    if module == "__builtin__":
```
Because `__builtin__` would never be in `sys.modules` so the next error message that `module` not in `sys.modules` would be inaccurate :)

albanD replied:
Sounds good!
Maybe update the comment :D

mikaylagawarecki commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +193,87 @@ def load(self):
+                            class_type, torch.Tensor
+                        ):
+                            # class_type.method accesses should call PyInstanceMethod_Function
+                            # https://github.com/python/cpython/blob/a37b0932285b5e883b13a46ff2a32f15d7339894/Objects/classobject.c#L374  # noqa: B950
```
I was wondering whether the `class_type.method` access could ever be overridden to do bad things, seems like answer is no though(?)

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allow tensor subclasses types that 
(1) Do not override `__setstate__`, `__getattr__`, `__setattr__`, `__get__`, `__set__` or `__getattribute__` of `torch.Tensor` (`torch.Tensor` does not have a definition of `__getattr__`, `__get__` or `__set__` so we check that these are `None`)
(2) Use the generic `tp_alloc`
(3) Are in a module that *has been imported by the user*
to be pushed onto the stack by `GLOBAL` instructions.

Further, we only allow these types to be used as the second argument of `_rebuild_from_type_v2`

*Note that we use `inspect.getattr_static(sys.modules[module], name)` to get the class/function as this method claims to have no code execution.

The rationale for the 3 conditions above is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `getattr`, `Tensor.__setstate__` and the call to `as_subclass` as well as the call to `_set_obj_state` which calls `setattr`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.{mark/get/clear}_safe_globals` that enables user to allowlist globals they have deemed safe and manipulate this list (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allow tensor subclasses types that 
(1) Do not override `__setstate__`, `__getattr__`, `__setattr__`, `__get__`, `__set__` or `__getattribute__` of `torch.Tensor` (`torch.Tensor` does not have a definition of `__getattr__`, `__get__` or `__set__` so we check that these are `None`)
(2) Use the generic `tp_alloc`
(3) Are in a module that *has been imported by the user*
to be pushed onto the stack by `GLOBAL` instructions.

Further, we only allow these types to be used as the second argument of `_rebuild_from_type_v2`

*Note that we use `inspect.getattr_static(sys.modules[module], name)` to get the class/function as this method claims to have no code execution.

The rationale for the 3 conditions above is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `getattr`, `Tensor.__setstate__` and the call to `as_subclass` as well as the call to `_set_obj_state` which calls `setattr`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.{mark/get/clear}_safe_globals` that enables user to allowlist globals they have deemed safe and manipulate this list (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Allow tensor subclasses and add `torch.serialization.mark_safe_globals` that allows users to allowlist classes for `weights_only` load"

#### Conditions for allowlisting tensor subclasses
We allow tensor subclasses types that 
(1) Do not override `__setstate__`, `__getattr__`, `__setattr__`, `__get__`, `__set__` or `__getattribute__` of `torch.Tensor` (`torch.Tensor` does not have a definition of `__getattr__`, `__get__` or `__set__` so we check that these are `None`)
(2) Use the generic `tp_alloc`
(3) Are in a module that *has been imported by the user*
to be pushed onto the stack by `GLOBAL` instructions.

Further, we only allow these types to be used as the second argument of `_rebuild_from_type_v2`

*Note that we use `inspect.getattr_static(sys.modules[module], name)` to get the class/function as this method claims to have no code execution.

The rationale for the 3 conditions above is as follows:

The rebuild func provided by `Tensor.__reduce_ex__` is `torch._tensor._rebuild_from_type_v2`, which is defined as such (note the call to `getattr`, `Tensor.__setstate__` and the call to `as_subclass` as well as the call to `_set_obj_state` which calls `setattr`)

https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/_tensor.py#L57-L71

`as_subclass` is implemented with a call to `THPVariable_NewWithVar`

that will eventually call `tp_alloc` here
https://github.com/pytorch/pytorch/blob/4e66aaa01092ddc8822bbca315b673329c76f4cd/torch/csrc/autograd/python_variable.cpp#L2053


The `func` arg to `_rebuild_from_type_v2` for wrapper subclasses is `Tensor.rebuild_wrapper_subclass`, which will similarly call into `THPVariable_NewWithVar` and hit the above `tp_alloc`

**Note that we do not call `tp_init` or `tp_new` (i.e. `cls.__init__` or `cls.__new__`) when unpickling**


### How do we check something is a tensor subclass/constraints around imports

In order to check whether `bla` is a tensor subclass in the bytecode `GLOBAL module.name`, we need to do an `issubclass` check, which entails converting the global string to the appropriate type. We *do not* arbitrarily import modules but will perform this check as long as the given subclass (given by `module.name`) has already been imported by the user (i.e. `module in sys.modules` and `issubclass(getattr(sys[modules], name), torch.Tensor)` 

This PR also allowlisted  `torch._utils._rebuild_wrapper_subclass` and `torch.device` (used by `_rebuild_wrapper_subclass`) 

### API for allow listing 
This PR also added `torch.serialization.{mark/get/clear}_safe_globals` that enables user to allowlist globals they have deemed safe and manipulate this list (for example they could allowlist a tensor subclass with a custom `__setstate__` if they have checked that this is safe.

Next steps:
- Add testing and allowlist required classes for all in-core tensor subclasses (e.g. `DTensor`, `FakeTensor` etc.)




[ghstack-poisoned]

mikaylagawarecki commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +193,87 @@ def load(self):
+                        raise RuntimeError(f"Unsupported class {full_path}")
+                    elif module not in modules:
+                        raise RuntimeError(
+                            f"Found global `{full_path}` in the checkpoint but `{module}` was "
```
@albanD I updated the wording significantly, including the exact criteria that we use. 

Let me know if this seems sufficient or whether we still need more details around "trust" here

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki committed:
Update

[ghstack-poisoned]

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +198,107 @@ def load(self):
+                    self.append(_get_user_allowed_globals()[full_path])
                 else:
-                    raise RuntimeError(f"Unsupported class {full_path}")
+                    # The logic in this branch handles user-defined tensor subclasses.
```
```suggestion
                    # The logic in this branch handles user-defined tensor subclasses we can automatically allow and raises an error for anything that is not provably safe.
```

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +198,107 @@ def load(self):
+                    #     and `tp_alloc`
+                    #     The methods that we ban overriding were selected in a test-driven manner
+                    #     by overriding every callable method on a tensor subclass and determinining
+                    #     which might get called when unpickling.
```
```suggestion
                    #     which might get called during unpickling.
```

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +198,107 @@ def load(self):
+                    #     by overriding every callable method on a tensor subclass and determinining
+                    #     which might get called when unpickling.
+                    # Further, these classes are only ever allowed as argument 2 of
+                    # `torch._tensor._rebuild_from_type_v2`.
```
```suggestion
                    # `torch._tensor._rebuild_from_type_v2` as other use of the class could use methods we didn't audit.
```

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +198,107 @@ def load(self):
+                    # Further, these classes are only ever allowed as argument 2 of
+                    # `torch._tensor._rebuild_from_type_v2`.
+                    if module == "__builtin__":
+                        raise RuntimeError(f"Unsupported class {full_path}")
```
The unsupported global error below is much better as it points to `add_safe_globals`. Could we re-use it here?

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +198,107 @@ def load(self):
+                            f"Found GLOBAL `{full_path}` instruction in the pickle file but `{full_path}` was "
+                            f"not in the pre-defined list of allowed globals that are considered safe by the "
+                            "weights_only unpickler for rebuilding state_dicts. This is the expected behavior if "
+                            f"`{full_path}` is a user-defined tensor subclass not defined in the `torch` package. "
```
This code will be called for any class that is not allowed for which the parent module is not already imported right?

mikaylagawarecki replied:
you're right! edited a bit to reflect this case

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +198,107 @@ def load(self):
+                            f"If this is the case, we expect `{module}` to be present in `sys.modules` (i.e. it "
+                            "must be imported in the current environment), but this was not the case. "
+                            f"If you intend to unpickle a `{full_path}` please import `{name}` from `{module}`. "
+                            f"Note that having this imported will *only* allow the type `{full_path}` to be passed "
```
We should have a docs page that go over this and point to it in the end. 
This is ok temporarily while we're getting this landing page.

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +198,107 @@ def load(self):
+                            "as well as the implementation of `tp_alloc`."
+                        )
+                    else:
+                        class_type = getattr_static(modules[module], name)
```
Can we check what the error is if my module uses `__getattr__` to dynamically serve the object?
In case it is very obscure, we should catch it here and add context that we're using getattr_static and so things might be missing.

mikaylagawarecki replied:
Is this what you mean:

```python
from inspect import getattr_static
class SpoofModule(ModuleType):
    def __getattr__(self, name):
        if name == 'TestEmptySubclass':
            return TestEmptySubclass

spoof_mod = SpoofModule('bla')
print(getattr(spoof_mod, 'TestEmptySubclass'))  # <class '__main__.TestEmptySubclass'> 
print(getattr_static(spoof_mod, 'TestEmptySubclass'))  # AttributeError: TestEmptySubclass
```

I don't think this error is expressive enough that we can catch it 🤔  wdyt?

albanD replied:
I would say:
```
try:
  class_type = getattr_static(modules[module], name)
except AttributeError as e:
  raise AttributeError("Some context about getattr_state") from e
```

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +198,107 @@ def load(self):
+                                is not torch.Tensor.__getattribute__
+                            )
+                            try:
+                                custom_get = class_type.__get__ is not None  # type: ignore[attr-defined]
```
Why not `custom_get = getattr(class_type, "__get__", None) is not None` ?

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -162,10 +308,28 @@ def load(self):
+                arg_is_subclass_type = [
+                    i
+                    for i, arg in enumerate(args)
+                    if isinstance(arg, type) and arg in self.tensor_subclasses_found
```
```suggestion
                    if isinstance(arg, type) and issubclass(arg, torch.Tensor)
```
Why not this?

mikaylagawarecki replied:
I think we should never let type be on the unpickler's stack

I changed the strategy here to 
(1) `GLOBAL` finds the subclass, if it meets the criteria, it pushes the str "module.attr" onto the unpickler's stack and adds tensor_subclasses_found["module.attr"] = module.attr
(2) when executing REDUCE for `rebuild_from_type_v2`, we properly translate the str back to the actual type

Does that sound reasonable to you?

albanD commented on test/test_serialization.py:
```diff
@@ -4128,6 +4148,201 @@ def test_empty_class_serialization(self):
+        Tests import semantic for tensor subclass and the {add/get/clear}_safe_globals APIs
+        '''
+        # Needed to prevent UnboundLocalError: local variable 'TwoTensor' referenced before assignment
+        global TwoTensor
```
Huh why?

mikaylagawarecki replied:
This is weird to me as well, but seems to happen because TwoTensor is imported again later in the test, can minimize this to this will still throw the error

```python

def test_safe_globals_for_weights_only(self):
    '''
    Tests import semantic for tensor subclass and the {add/get/clear}_safe_globals APIs
    '''
    t = TwoTensor(torch.randn(2, 3), torch.randn(2, 3))
    from torch.testing._internal.two_tensor import TwoTensor
 ```

raises 
```
t = TwoTensor(torch.randn(2, 3), torch.randn(2, 3))
UnboundLocalError: local variable 'TwoTensor' referenced before assignment
```

albanD reviewed (commented):
Sounds good!

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki committed:
Update

[ghstack-poisoned]

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -151,8 +198,124 @@ def load(self):
+                            class_type = getattr_static(modules[module], name)
+                        except AttributeError as e:
+                            raise AttributeError(
+                                f"For safety, we use inspect.getattr_state to get {name} from {module} "
```
```suggestion
                                f"For safety during weights_only loading, we use inspect.getattr_state to get {name} from {module} "
```

albanD reviewed (approved):
Sounds good!

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
